sympy/polys/series/python_powerseriesring.py

```diff
@@ -0,0 +1,799 @@
+        return s
+    if n == -1:
+        return _useries_neg(s, dom, ring_prec)
+
+    series = dup_mul_ground(coeffs, n, dom)
+    return _useries(series, prec, dom, ring_prec)
+
+
+def _useries_div(
+    s1: USeries[Er], s2: USeries[Er], dom: Domain, ring_prec: int
+) -> USeries[Er]:
+    coeffs1, prec1 = s1
+    coeffs2, prec2 = s2
+
+    if not coeffs2:
+        raise ZeroDivisionError("Series division by zero")
+
+    val1 = _useries_valuation(s1, dom)
+    val2 = _useries_valuation(s2, dom)
+
+    if val1 < val2:
+        raise ValueError("quotient would not be a power series")
+
+    if not dom.is_unit(coeffs2[-val2 - 1]):
+        raise ValueError("Trailing coefficient of the divisor must be a unit")
+
+    if val2 == 0:
+        _, _, prec = _unify_prec(s1, s2, dom, ring_prec)
+        s2_inv = _useries_inverse(s2, dom, prec)
+        return _useries_mul(s1, s2_inv, dom, prec)
+    else:
+        # Shift both series to make divisor's valuation zero
+        shifted_coeffs1 = dup_rshift(coeffs1, val2, dom)
+        shifted_coeffs2 = dup_rshift(coeffs2, val2, dom)
+
+        cap = ring_prec - val2
+
+        prec1 = (prec1 if prec1 is not None else ring_prec) - val2
+        prec2 = (prec2 if prec2 is not None else ring_prec) - val2
+
+        shifted_s1 = _useries(shifted_coeffs1, prec1, dom, cap)
+        shifted_s2 = _useries(shifted_coeffs2, prec2, dom, cap)
+
+        q_coeffs = _useries_div(shifted_s1, shifted_s2, dom, cap)[0]
+
+        _, _, prec = _unify_prec(shifted_s1, shifted_s2, dom, ring_prec)
+        return q_coeffs, prec
+
+
+def _useries_pow_int(
+    s: USeries[Er], n: int, dom: Domain, ring_prec: int
+) -> USeries[Er]:
+    """Raise a power series to a non-negative integer power with truncation."""
+    if n < 0:
+        raise ValueError("Power must be a non-negative integer")
+
+    coeffs, prec = s
+
+    if n == 0:
+        return [dom.one], prec
+
+    if prec is None:
+        deg = dup_degree(coeffs) * n
+        if deg < ring_prec:
+            return dup_series_pow(coeffs, n, ring_prec, dom), None
+        prec = ring_prec
+
+    series = dup_series_pow(coeffs, n, prec, dom)
+    return series, prec
+
+
+def _useries_truncate(s: USeries[Er], n: int, dom: Domain) -> USeries[Er]:
+    """Truncate a power series to the first n terms."""
+    coeffs, prec = s
+
+    if n < 0:
+        raise ValueError("Truncation precision must be non-negative")
+
+    deg = dup_degree(coeffs)
+    if deg < n:
+        return coeffs, prec
+    return dup_truncate(coeffs, n, dom), n
+
+
+def _useries_compose(
+    s1: USeries[Er], s2: USeries[Er], dom: Domain, ring_prec: int
+) -> USeries[Er]:
+    """Compose two power series."""
+    coeffs1, prec1 = s1
+    coeffs2, prec2 = s2
+
+    if prec1 is None and prec2 is None:
+        comp = dup_series_compose(coeffs1, coeffs2, ring_prec, dom)
+
+        deg1 = dup_degree(coeffs1)
+        deg2 = dup_degree(coeffs2)
+        if deg1 + deg2 < ring_prec:
+            return comp, None
```

Comment:
Should this not check the degree of `comp`?

I think that the degrees multiply rather than add:
```python
In [15]: p3 = flint.fmpz_poly([0, 1, 2, 3])

In [16]: p2 = flint.fmpz_poly([1, 2, 3])

In [17]: p2
Out[17]: 3*x^2 + 2*x + 1

In [18]: p3
Out[18]: 3*x^3 + 2*x^2 + x

In [19]: p2(p3)
Out[19]: 27*x^6 + 36*x^5 + 30*x^4 + 18*x^3 + 7*x^2 + 2*x + 1
```